test/hashing.jl

```diff
@@ -72,25 +73,77 @@ vals = Any[
     zeros(2, 2), spzeros(2, 2), eye(2, 2), speye(2, 2),
     sparse(ones(2, 2)), ones(2, 2), sparse([0 0; 1 0]), [0 0; 1 0],
-    [-0. 0; -0. 0.], SparseMatrixCSC(2, 2, [1, 3, 3], [1, 2], [-0., -0.])
+    [-0. 0; -0. 0.], SparseMatrixCSC(2, 2, [1, 3, 3], [1, 2], [-0., -0.]),
+    # issue #16364
+    1:4, 1:1:4, 1:-1:0, 1.0:4.0, 1.0:1.0:4.0, linspace(1, 4, 4),
+    'a':'e', ['a', 'b', 'c', 'd', 'e'],
+    # check that hash is still consistent with heteregeneous arrays for which - is defined
+    # for some pairs and not others
+    ["a", "b", 1, 2], ["a", 1, 2], ["a", "b", 2, 2], ["a", "a", 1, 2], ["a", "b", 2, 3]
 ]
 
 for a in vals, b in vals
     @test isequal(a,b) == (hash(a)==hash(b))
 end
 
+vals = Any[
+    Int[], Float64[],
+    [0], [1], [2],
+    # test various sparsity patterns with repetitions of steps
+    [0, 0], [0, 0, 0], [0, 1], [1, 0],
+    [0, 0, 1], [0, 1, 0], [1, 0, 0], [0, 1, 2],
+    [0 0; 0 0], [1 0; 0 0], [0 1; 0 0], [0 0; 1 0], [0 0; 0 1],
+    [5 1; 0 0], [1 1; 0 1], [0 2; 3 0], [0 2; 4 6], [4 0; 0 1],
+    [0 0 0; 0 0 0], [1 0 0; 0 0 1], [0 0 2; 3 0 0], [0 0 7; 6 1 2],
+    [4 0 0; 3 0 1], [0 2 4; 6 0 0],
+    # run of equal steps that crosses a zero
+    [0 3 2 1 0 -1], [0 1 0 -1 0],
+    # various stored zeros patterns
+    sparse([1], [1], [0]), sparse([1], [1], [-0.0]),
+    sparse([1, 2], [1, 1], [-0.0, 0.0]), sparse([1, 2], [1, 1], [0.0, -0.0]),
+    sparse([1, 2], [1, 1], [-0.0, 0.0], 3, 1), sparse([1, 2], [1, 1], [0.0, -0.0], 3, 1),
+    sparse([1, 3], [1, 1], [-0.0, 0.0], 3, 1), sparse([1, 3], [1, 1], [0.0, -0.0], 3, 1),
+    sparse([1, 2, 3], [1, 1, 1], [-1, 0, 1], 3, 1), sparse([1, 2, 3], [1, 1, 1], [-1.0, -0.0, 1.0], 3, 1),
+    sparse([1, 3], [1, 1], [-1, 0], 3, 1), sparse([1, 2], [1, 1], [-1, 0], 3, 1)
+]
+
+for a in vals
+    b = Array(a)
+    @test hash(convert(Array{Any}, a)) == hash(b)
+    @test hash(convert(Array{supertype(eltype(a))}, a)) == hash(b)
+    @test hash(sparse(a)) == hash(b)
+end
+
+# Test hashing sparse matrix with type which does not support -
+struct CustomHashReal
+    x::Float64
+end
+Base.hash(x::CustomHashReal, h::UInt) = hash(x.x, h)
+Base.:(==)(x::CustomHashReal, y::Number) = x.x == y
+Base.:(==)(x::Number, y::CustomHashReal) = x == y.x
+Base.zero(::Type{CustomHashReal}) = CustomHashReal(0.0)
+
+let a = sparse([CustomHashReal(0), CustomHashReal(3), CustomHashReal(3)])
+    @test hash(a) == hash(Array(a))
+end
+
+vals = Any[
+    0.0:0.1:0.3, 0.3:-0.1:0.0,
+    0:-1:1, 0.0:-1.0:1.0, 0.0:1.1:10.0, -4:10,
+    'a':'e', 'b':'a',
+    linspace(1, 1, 1), linspace(0.3, 1.0, 3),  linspace(1, 1.1, 20)
+]
+
+for a in vals
+    @test hash(collect(a)) == hash(a)
+end
+
 @test hash(SubString("--hello--",3,7)) == hash("hello")
 @test hash(:(X.x)) == hash(:(X.x))
 @test hash(:(X.x)) != hash(:(X.y))
 
 @test hash([1,2]) == hash(view([1,2,3,4],1:2))
 
-# test explicit zeros in SparseMatrixCSC
-x = sprand(10, 10, 0.5)
-x[1] = 1
-x.nzval[1] = 0
-@test hash(x) == hash(Array(x))
```

Comment:
The new tests are much more systematic and cover this particular case of a stored zero. Other than that, it's not clear what this code is testing since it uses random values. Better only keep tests that have a clear meaning -- if something is missing I'd rather add it to the loop above.